fix(wal): ensure WAL offset monotonicity under S3 backpressure

Chandi977 opened the pull request:
## Summary

This PR fixes a correctness issue in the S3 WAL implementation where **multiple concurrent bulk uploads may reuse the same `baseOffset`** when S3 backpressure or slow storage causes delayed uploads.  
As a result, two or more bulks can start from the **same offset**, producing **duplicate WAL offsets** and violating the monotonicity guarantees required for Kafka-compatible behavior.

The root cause is that `nextOffset` was only advanced *after* bulk upload scheduling, allowing concurrent bulks to observe identical starting offsets.

---

## Fix

This PR introduces **atomic offset claiming** using a lightweight lock:

- Added `offsetClaimLock`
- Added `claimOffsetRange()` which:
  - Atomically assigns offsets for every record in a bulk
  - Advances `nextOffset` before any upload or buffer creation starts
  - Prevents duplicate offsets even under severe backpressure

The change is isolated and does **not modify the bulk batching logic**, only the correctness of offset assignment.

---

## Changes Included

- Updated `DefaultWriter.java`:
  - Added `offsetClaimLock`
  - Added `claimOffsetRange()` helper
  - Updated `uploadBulk0()` to use atomic offset assignment
  - Ensures `endOffset` aligns correctly via `ceilAlignOffset`
- No changes to interfaces or external behavior

---

## Reproduction (Before Fix)

1. Run two high-throughput producers.
2. Pause or slow MinIO (`docker pause minio`).
3. Inspect WAL logs:  
   Overlapping or repeated offsets are produced during concurrent uploads.

---

## Validation (After Fix)

- No duplicate offsets observed under backpressure.
- Offsets strictly increase across all streams.
- WAL recovery produces consistent results.
- Concurrency test + stress test run successfully.

---

## Issue Reference

Fixes: #3033  
**WAL offset monotonicity violation under S3 backpressure**

---

## Impact

✔ Guarantees monotonic offset progression  
✔ Prevents WAL corruption under concurrency  
✔ Safe for existing deployments  
✔ Zero performance regression (offset lock is lightweight)  

---

## Notes for Reviewers

This patch focuses solely on correctness.  
No behavior changes in batching, uploads, or IO.  
Only offset assignment is made deterministic and atomic.

Happy to adjust tests or add benchmarks if required.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/AutoMQ/automq?pullRequest=3034) <br/>All committers have signed the CLA.

Chandi977 commented:
Hi @superhx — thanks for the review.  
Below is a concise, technical summary you can use to validate the issue and the fix included in this PR.

---

## Summary
**Issue:** Under S3 upload backpressure, concurrent bulks can claim the same `nextOffset`, causing overlapping WAL object ranges (duplicate/overlapping offsets).  
**Root cause:** Offset claiming and offset advancement are not atomic — two threads can read the same `nextOffset` before it is advanced. The existing `writeLock` does not prevent this race because claiming occurs outside the code path that advances the global offset when uploads are slow.  
**Fix:** Make the offset claim-and-advance atomic (short critical section) so a bulk claims a contiguous offset range and immediately advances `nextOffset` before the upload prepares buffers.

---

## Reproduction (high level)
1. Start S3-compatible storage (e.g. MinIO).  
2. Run two concurrent producers that create and enqueue bulks in rapid succession.  
3. Introduce short S3 slowdown / backpressure (pause/unpause container or throttle network) so multiple bulks are prepared before uploads complete.  
4. Observe two bulks assigned the same base offset and resulting overlapping WAL objects.

---

## Minimal technical explanation & timeline

```

T1: read nextOffset = 2000  ───────┐
│  (S3 slow)
T2: read nextOffset = 2000  ───────┘

T1: assign baseOffset = 2000
T2: assign baseOffset = 2000   <-- collision

T1: uploadActiveBulk -> nextOffset += 400  (too late)
T2: uploadActiveBulk -> nextOffset += 500

```

Because both threads read the same `nextOffset` before either advances it, both assign overlapping ranges. This is a classic race: reads are concurrent and the advance happens later during upload scheduling.

---

## What I changed in this PR
- **Atomic offset claim**: I added a short critical section that:
  - computes per-record offsets and assigns them immediately, and
  - advances `nextOffset` immediately at claim time.
- This ensures that no two bulks/records can observe the same `nextOffset` and avoids overlapping WAL objects regardless of S3 latency.

**Key property:** the critical section is short (only offset arithmetic and assignment) — it does not perform IO or buffer allocation, so it does not hurt throughput but prevents offset collisions under backpressure.

---

## Why this is safe and correct
- Preserves *monotonicity* of offsets (strictly increasing global `nextOffset`).
- Keeps uploads and heavy IO outside the critical section to avoid contention.
- Prevents WAL recovery ambiguity caused by overlapping object offsets.

---

## Observability (what to check after applying PR)
- Under the same high-load/backpressure reproduction, confirm:
  - No two bulks share the same `baseOffset`.
  - WAL objects in storage have contiguous, non-overlapping `[startOffset,endOffset)` ranges.
  - `nextOffset` always increases monotonically even when uploads are slow.

---

## Example WAL inspection (what to look for)
- **Bad (before fix):**
```

Object A -> [2000, 2400)
Object B -> [2000, 2500)   <-- overlap (bad)

```
- **Good (after fix):**
```

Object A -> [2000, 2400)
Object B -> [2400, 2900)   <-- contiguous, non-overlapping (good)

```

---

## Additional notes
- The fix touches only the offset-claiming logic; upload flow and retry semantics are unchanged.
- The atomic claim is intentionally narrow in scope to avoid performance regressions.
- I validated the behavior by forcing an upload slowdown and verifying offsets no longer collide.
 
Thanks.

Chandi977 commented:
Hi @Rancho-7, thanks for the follow-up!

You're correct that `uploadBulk0()` itself is single-threaded per bulk.
But the issue happens **before upload**, during **bulk creation**, where `baseOffset` is assigned.

The race looks like this:

```
T1: append(...) 
 → reads nextOffset = 2000
 → builds bulk B1 (baseOffset = 2000)
 → S3 slow → uploadActiveBulk() delayed

T2: append(...)
 → reads nextOffset = 2000   (still unchanged)
 → builds bulk B2 (baseOffset = 2000)
```

Both bulks now start at **the same baseOffset**, even though they contain different data.

The key point is:

* `nextOffset` is advanced **only inside `uploadActiveBulk()`**
* `append()` can be called by multiple threads **before the offset advance happens**
* There is no lock that makes
  `(claim nextOffset → assign baseOffset)`
  an atomic operation

So although upload is serialized, **bulk creation is not**, and two threads can observe the same `nextOffset` before it is moved forward.

The fix adds a small lock ensuring:

```
claim nextOffset
assign baseOffset
advance nextOffset
```

happen as one atomic step, preventing duplicate WAL offsets under S3 backpressure.
